File: packages/android-sdk/src/main/kotlin/com/phosra/sdk/networking/ApiError.kt
package com.phosra.sdk.networking

/**
 * Sealed class hierarchy for all API errors the SDK can produce.
 * Consumers can exhaustively match on these to handle every error case.
 */
sealed class PhosraApiError : Exception() {

    /** HTTP error with a non-success status code. */
    data class HttpError(
        val statusCode: Int,
        override val message: String?
    ) : PhosraApiError()

    /** Network-level failure (no connectivity, DNS resolution, timeout, etc.). */
    data class NetworkError(
        override val cause: Throwable
    ) : PhosraApiError() {
        override val message: String = cause.message ?: "Network error"
    }

    /** Failed to decode the JSON response body. */
    data class DecodingError(
        override val cause: Throwable
    ) : PhosraApiError() {
        override val message: String = cause.message ?: "JSON decoding error"
    }

    /** 401 Unauthorized — device key or bearer token is invalid/expired. */
    data object Unauthorized : PhosraApiError() {
        override val message: String = "Unauthorized"
    }

    /** 404 Not Found. */
    data object NotFound : PhosraApiError() {
        override val message: String = "Not found"
    }

    /** 429 Too Many Requests — includes optional Retry-After header value. */
    data class RateLimited(
        val retryAfterSeconds: Long?
    ) : PhosraApiError() {
        override val message: String = "Rate limited" +
            (retryAfterSeconds?.let { ", retry after ${it}s" } ?: "")
    }

    /** 304 Not Modified — policy has not changed since the requested version. */
    data object NotModified : PhosraApiError() {
        override val message: String = "Not modified"
    }

    /** No device key is stored — must register the device first. */
    data object NoDeviceKey : PhosraApiError() {
        override val message: String = "No device key available"
    }
}

File: packages/ios-sdk/Sources/PhosraSDK/Networking/APIError.swift
import Foundation

/// Errors that can occur when communicating with the Phosra API.
public enum PhosraAPIError: Error, LocalizedError {
    /// A network-level error (no connectivity, DNS failure, timeout, etc.).
    case networkError(Error)

    /// The server returned an HTTP error status code.
    case httpError(statusCode: Int, message: String?)

    /// The response body could not be decoded into the expected type.
    case decodingError(Error)

    /// 401 Unauthorized -- the device key or parent token is invalid or expired.
    case unauthorized

    /// 404 Not Found -- the requested resource does not exist.
    case notFound

    /// 429 Too Many Requests -- the client is being rate-limited.
    case rateLimited(retryAfter: TimeInterval?)

    /// 304 Not Modified -- the policy has not changed since the requested version.
    case notModified

    /// No device key is configured. Required for device-authenticated endpoints.
    case noDeviceKey

    /// No parent token is configured. Required for parent-authenticated endpoints.
    case noParentToken

    public var errorDescription: String? {
        switch self {
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .httpError(let statusCode, let message):
            if let message = message {
                return "HTTP \(statusCode): \(message)"
            }
            return "HTTP error \(statusCode)"
        case .decodingError(let error):
            return "Decoding error: \(error.localizedDescription)"
        case .unauthorized:
            return "Unauthorized: invalid or expired credentials"
        case .notFound:
            return "Resource not found"
        case .rateLimited(let retryAfter):
            if let retryAfter = retryAfter {
                return "Rate limited. Retry after \(Int(retryAfter)) seconds."
            }
            return "Rate limited. Please try again later."
        case .notModified:
            return "Not modified (304)"
        case .noDeviceKey:
            return "No device key configured. Call PhosraClient.forDevice() or register a device first."
        case .noParentToken:
            return "No parent token configured. Call PhosraClient.forParent() first."
        }
    }
}
